[smt] smt_dump is no longer an error for Text

rafaelsamenezes opened the pull request:
Running --smtlib --output gives an error because smt_dump is not supported by the interface. ESBMC, however, requires all solvers to have it.

rafaelsamenezes committed:
[smt] smt_dump is no longer an error for Text

fbrausse commented on src/solvers/smt/smt_conv.cpp:
```diff
@@ -2728,8 +2728,7 @@ smt_astt smt_ast::project(
 {
-  log_error("SMT dump not implemented for {}", solver_text());
-  abort();
+  log_status("SMT dump was sent to --output file");  
```
If this function is not overridden and executed, printing the message would factually be wrong since `dump_smt()` did not produce any output to the file. Assuming that every backend is supposed to implement `dump_smt()`, wouldn't the proper solution be to implement it for the smtlib backend? Am I missing something?

rafaelsamenezes replied:
So just dump the formula at stdout?

fbrausse replied:
That's what I would expect. *Edit:* unless `--output` was given. But again, the output management is a bit unclear right now since ESBMC can generate lots of outputs.

rafaelsamenezes replied:
Yeah, the `--output` is a very misleading flag. Also, its extra confusing for smtlib output.

For other solvers, using `--smt-formula-<only/too>` means to invoke the `dump_smt()` method. For the smtlib, generating a file/output is a requirement. So its redundant to print it out again.

fbrausse replied:
> For the smtlib, generating a file/output is a requirement. So its redundant to print it out again.

It only is required if no `--smtlib-solver-prog` is given. If the solver-prog is used, based just on the help message description I would expect `--output` to still write whatever formula the smtlib backend produces to the given file, however, it refuses: `ERROR: Can't solve SMTLIB output and write to a file, sorry`.

So on a non-Windows build right now, `smtlib` requires either the `--output` or the `--smtlib-solver-prog` option to be present. Both options result in the above error, none of them are `ERROR: Must specify an smtlib solver program in smtlib mode`. Windows builds don't support `--smtlib-solver-prog` because process spawning is not implemented. Are you referring to and is this PR fixing the Windows build?

As an aside, `--smtlib-solver-prog` does not support handling spaces or command-line parameters to the executable, which could be fixed by exec'ing `sh -c PROG`.

rafaelsamenezes replied:
At this point, I would say that the correct solution would be:

1. Ignore --output at convertion and only use it while dumping the formula.
2. This PR do not aim to fix the windows build, it's just enabling the usage of --smtlib to output formulas.

> As an aside, --smtlib-solver-prog does not support handling spaces or command-line parameters to the executable, which could be fixed by exec'ing sh -c PR

Could you clarify or point a link? I am having trouble finding the man page describing what PR does.

fbrausse replied:
> > As an aside, --smtlib-solver-prog does not support handling spaces or command-line parameters to the executable, which could be fixed by exec'ing sh -c PR
> 
> Could you clarify or point a link? I am having trouble finding the man page describing what PR does.

PROG in `sh -c PROG` was meant to stand for the parameter given to `--smtlib-solver-prog`. For instance, if I want to use Z3 as the external solver, I would like/need to give `z3` the parameter `-in` in order for it to read from `stdin`. But specifying `--smtlib-solver-prog "z3 -in"` does not work as the program exec'ed by the smtlib backend now is `z3 -in` as one string. Such an executable does not exist. For comparison, this is similar to what was done in https://github.com/esbmc/esbmc/pull/602/files#diff-6d77e207194f76d3f2d17e0ef4b551cbf8bcc60848012dfb4cdaba0a8efae610 with the `ESBMC_OPTS` envvar: an arbitrary string that gets interpreted as a sequence of separate options by some simple rules. The rules in the `sh -c PROG` example would be the shell-parsing rules. For `ESBMC_OPTS` I've added a small method that does the argument splitting on a reduced subset of those shell-parsing rules (e.g., basic `\`-escaping, unescaped white-space separates options, quoting via `"`). That method could either be re-used for consistency or we could just delegate to the shell.

fbrausse replied:
> At this point, I would say that the correct solution would be:
>  1. Ignore --output at convertion and only use it while dumping the formula.

Agreed.

fbrausse replied:
> As an aside, --smtlib-solver-prog does not support handling spaces or command-line parameters to the executable, which could be fixed by exec'ing sh -c PR

I'll submit a PR that does this after this one here is merged.

rafaelsamenezes committed:
[smtlib] smtlib now overrides dump_smt

rafaelsamenezes committed:
[smtlib] smtlib_smt_sort is now used

Prevents a segmentation fault when converting arrays to text as the api expects
it to be of type smtlib_smt_sort.

rafaelsamenezes committed:
[smtlib] mk_concat now has a bigger sort

rafaelsamenezes committed:
[testing] testing tool is now able to test whether the program was able to produce a formula

fbrausse commented on src/solvers/smtlib/smtlib_conv.cpp:
```diff
@@ -104,12 +104,38 @@ smt_convt *create_new_smtlib_solver(
+    log_error(
+      "[smtlib] the smtlib interface does not support solving for now. Please, "
+      "use it with --smt-formula-only");
+    abort();
```
Would it be possible to make this a warning instead? The `--smtlib-solver-prog` option still works even though parsing the replies from the external program is somewhat buggy. Making this a warning would help in debugging it.

rafaelsamenezes replied:
Sure

fbrausse reviewed (approved):
Generally looks good to me, just one question.

rafaelsamenezes committed:
[smtlib] smt-formula-only is no longer mandatory

rafaelsamenezes commented:
There is still a couple of issues and inconssistencies that need to be sort out (mainly shown during the inductive step). They will come in other PRs though.
